src/ConfigSpace/_condition_tree.py

```diff
@@ -0,0 +1,698 @@
+        return "\n".join(parts)
+
+
+@dataclass
+class HPNode:
+    hp: Hyperparameter
+    idx: int
+    maximum_depth: int
+    children: dict[str, tuple[HPNode, ConditionLike]] = field(default_factory=dict)
+    forbiddens: list[ForbiddenLike] = field(default_factory=list)
+
+    # NOTE: We have the restriction that a hyperparameter can only have one parent
+    # condition but multiple parent from which these relationshops are dervied
+    # This is to prevent ambiguity between AND and OR, in other words, what
+    # do we do with the current Node if the condition for one parent is satisfied
+    # but not the other.
+    # * A useful assertion to make is that all nodes in `parents` will have the same
+    #   parent condition, also accessible via `parent_condition`
+    parents: dict[str, tuple[HPNode, ConditionLike]] = field(
+        default_factory=dict,
+        # We explicitly don't compare parents to prevent recursion
+        compare=False,
+    )
+    parent_condition: ConditionLike | None = None
+
+    def __lt__(self, __value: object) -> bool:
+        if not isinstance(__value, HPNode):
+            return NotImplemented
+
+        return (self.maximum_depth, self.hp.name) < (
+            __value.maximum_depth,
+            __value.hp.name,
+        )
+
+    def __le__(self, __value: object) -> bool:
+        if not isinstance(__value, HPNode):
+            return NotImplemented
+
+        return (self.maximum_depth, self.hp.name) <= (
+            __value.maximum_depth,
+            __value.hp.name,
+        )
+
+    def __gt__(self, __value: object) -> bool:
+        if not isinstance(__value, HPNode):
+            return NotImplemented
+
+        return (self.maximum_depth, self.hp.name) > (
+            __value.maximum_depth,
+            __value.hp.name,
+        )
+
+    def __ge__(self, __value: object) -> bool:
+        if not isinstance(__value, HPNode):
+            return NotImplemented
+
+        return (self.maximum_depth, self.hp.name) >= (
+            __value.maximum_depth,
+            __value.hp.name,
+        )
+
+    @property
+    def name(self) -> str:
+        return self.hp.name
+
+    def child_conditions(self) -> list[Condition | Conjunction]:
+        return [condition for _, condition in self.children.values()]
+
+    def parent_conditions(self) -> list[Condition | Conjunction]:
+        if self.parent_condition is not None:
+            return [self.parent_condition]
+        return []
+
+    def propogate_new_depth(self, depth: int) -> None:
+        if depth <= self.maximum_depth:
+            # There's no need to propogate the depth, this nodes
+            # maximum depth already dominates it
+            return
+
+        self.maximum_depth = depth
+        for child, _ in self.children.values():
+            child.propogate_new_depth(depth + 1)
+
+
+@dataclass
+class DAG:
+    # All relevant information is kept in these fields, rest is cached information
```

Comment:
Could you please make this comment more specific? Right now it is not clear to which variables you mean by "these fields". I would have expected "keep track of conditions" is also relevant, because the conditions also the vertices of the graph.

Reply:
Yup, I will have to take time for all the documentation!